Handle unknown commands / generators

kevingimbel opened the pull request:
This commit adds a `validateGenerators` function to the main loop which
validates the passed generators and stops execution if a validator is
not found.

While starting my work on #9 I realized that it is possible that one or more generators / arguments can be wrong and therefore result in unexpected output. For example, running the following command will create a "wrong" output. 

```sh
$ fakedata email name.first name.lst state country --format csv
```

The parameter `name.lst` is spelled wrong but previously `fakedata` did not throw errors so it would simply create a file in the format `email, name.first, state, country`. With this commit and the new `validateGenerators` function it throws an error:

```sh
$ fakedata email name.first name.lst state country --format csv

  Unknown generator: name.lst.

  Please check fakedata --generators for more information about available generators.
```

I am not entirely happy with the implementation of `validateGenerators`. Because I did not find a way to get the generators map I am creating a temporary one inside `validateGenerators` to "easily" check if any of the supplied generators is not valid. I did some (poor) performance testing by running `fakedata` with and without `validateGenerators` but could not observe any impacts.

lucapette commented:
First of all, let me thank you one more time for taking some time and work on fakedata. It's great!

I'll try to give you as much feedback as possible because you told me you're new to Go so this is a learning experience. I always feel the need to underline best intentions because getting _a lot_ of feedback can be a bit overwhelming. 

As I wouldn't want to merge a PR that adds new behaviour without some tests, my first suggestion is to move the function you wrote inside the package. I think `pkg/fakedata/fakedata.go` would do it for now. That has a few consequences:

- there's no need to pass in the generator list anymore because, since you're in the package, there's a [generators map available](https://github.com/lucapette/fakedata/blob/master/pkg/fakedata/generator.go#L20). Which is probably easier to use too. It's a map so we'd only need to check if the key is there or not. 
- I wouldn't access `flag.Args` from inside the package so I think a good way of solving the problem is to pass in the a `keys []string` to the new function
- the function needs to be public (that means it needs to start with a capital letter in Go) otherwise we can't use it from `cmd/fakedata.go`. 
- I wouldn't print inside the new function. In Go, if something goes wrong you return an error from the function. It's a feature I like a lot despite the noise all the `if err !=nil` create in the code

Putting all these things together means creating a function with a signature that goes like:

```go
func ValidateGenerators(keys []string) error
```

That way we can create a new test function in `pkg/fakedata/fakedata_test.go` that looks like this:


```go
func TestValidateGenerators(t *testing.T) {
	type args struct {
		keys []string
	}
	tests := []struct {
		name    string
		args    args
		wantErr bool
	}{
	// TODO: Add test cases.
         {"known generator", []string{"email"}, false},
         {"unknown generator", []string{"theresnosuchgenerator"}, true},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if err := fakedata.ValidateKeys(tt.args.keys); (err != nil) != tt.wantErr {
				t.Errorf("ValidateKeys() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}
```
This test code is untested (how ironic, right?). This strange `tests` variable is a common Go technique for testing I enjoy a lot. They are called [Table Driven Tests](https://github.com/golang/go/wiki/TableDrivenTests) and they're a common (and I'd argue) good practise of the Go community. It's easy to extend a feature or add new tests cases (that cover a bug fix for example) with this technique.

I hope this is helpful! Please let me know what you think and you'd like to change the code in the direction I suggested.

kevingimbel commented:
Thanks for your feedback! I agree that moving the function to `pkg/fakedata/fakedata.go` is a good idea and in fact the "right way" to handle it. I'll update the PR as soon as I can work on it.
I'll also add the tests, it bothered me that I couldn't add them before because I - like you - don't like to not have tests, especially when the language has testing build in. I'll write some tests and add them to the `fakedata_test.go` file.

Since the direction changed a bit, do you like the new behaviour? Previously fakedata would generate a dataset regardless of all parameters being a valid generator, not it stops and prints the error. Maybe we could also log an issue instead but generate the data anyway? I guess this depends on how we (or you) want to treat a falsy generator. Is it an error or is it a warning? I'd say it's an error and users should be encouraged to only use valid generators.

lucapette commented:
@kevingimbel I think the previous behavior was worse than what you suggest, especially if you generate a large number of columns/rows. It can be pretty annoying to have an empty column and only realising it after a while. So I'm with you saying that we should treat it as an error.

kevingimbel commented:
I moved the `ValidateGenerators` function into `pkg/fakedata/fakedata.go` and re-wrote it so it uses the existing `generators` map (https://github.com/kevingimbel/fakedata/blob/c8ee6d0e06430ccbca9a096a04c85f585f9eb6bc/pkg/fakedata/fakedata.go#L30-L49). Temporary I had to change the import to `github.com/kevingimbel/fakedata` so my new, changed `fakedata.go` file was included.

I then added a test based on your pseudo test code which worked very well. I added 6 cases for different combinations of existing and non-existing generators (https://github.com/lucapette/fakedata/pull/10/commits/ea8808aef60f7f24a6eabb530eaf158e3f89d23e). 

As you can see all tests pass and Travis is green, yay! 🎉

The new behaviour stops the execution whenever one or more not existing generators are found and then prints one or more error messages.

```sh
$ ./fakedata name.first name.last email city state country --limit 1

  Unknown generator: city.

  See fakedata --generators for a list of all available generators.
```

```sh
$ ./fakedata first last email salery,50k..300k password

  Unknown generator: first.
  Unknown generator: last.
  Unknown generator: salery.
  Unknown generator: password.

  See fakedata --generators for a list of all available generators.
```

lucapette commented:
@kevingimbel thank you very much. I just merged it 🎉 

I did some minor tweaks so that the code is a bit more idiomatic Go. You can see them [here](https://github.com/lucapette/fakedata/commit/1075fd39a33798f4b38a55e57d59b2b587b8e6d3), I tried to explain all of them in the commit message. I hope that helps! Please let me know if you have any questions, it was great to have your contribution!

kevingimbel commented:
@lucapette thanks! Looking at the code most of the changes make sense and are understandable to me, however I have two questions about the error handling. 

By adding `(err error)` to the `ValidateGenerators` function return did you initialize an "empty" `err`? At first I was trying to create a `nil` variable to return but apparently that doesn't work in Go. 

In `cmd/fakedata.go` you handle the error "inline" like so:
```go
if err := fakedata.ValidateGenerators(flag.Args()); err != nil {
	fmt.Printf("%v\n\nSee fakedata --generators for a list of available generators.\n", err)
	os.Exit(0)
}
```

Is this just a pattern in Go or is there any advantage compared to writing it in two statements, e.g.
```go
err := fakedata.ValidateGenerators(flag.Args())
if err != nil {
	fmt.Printf("%v\n\nSee fakedata --generators for a list of available generators.\n", err)
	os.Exit(0)
}
```

Looking at the two side by side I'd prefer to first one because it encapsulates the handling and execution into a "logical block", but are there any other reasons? Or just personal preferences? 

Thanks for the great code review, I'm happy to contribute again! :)

lucapette commented:
@kevingimbel 

`(err error)` follows the [zero value](https://tour.golang.org/basics/12) rule as it's a variable declaration. what you wrote initially is indeed a way of declaring it nil `var err error`. What I did is just a bit shorter and it makes the code much more explicit (imo)

About the if in the main function, the version you like the most is more common and I generally prefer that version too. In this case, I specifically inlined it because err would be used only inside the if (there's no other error handling in the entire main.. strange given it's go :D). In a way, you can say it's a personal preference
